docs/rfcs/0004-agent-flow.md

```diff
@@ -0,0 +1,162 @@
+
+# 2. Broad Solution Path
+
+Conversation around should the components be assembled via message passing, via expressions, or a hybrid approach.
+
+**Note: Consider all examples pseudoconfig**
+
+## 2.1 Expression Based
+
+Expression based is writing expressions that allow referencing other components streams/outputs/values and using them directly. 
+
+**Pros**
+
+* Easier to Implement, evaluating expressions can map directly to existing config structs
+* Components are more reusable, you can pass basic types arounds (string, int, bool)
+
+**Cons**
+* Harder for users to wire things together
+* Harder to build a GUI for
+
+
+## 2.2 Message Based
+
+Message based is where components have no knowledge of other components and information is passed strictly via input and output streams. 
+
+**Pros**
+
+* Easier for users to understand the dependencies between components
+* Easier to build a GUI for
+* Ability to use more configuration formats easier
+
+**Cons**
+
+* More time consuming to implement, existing integrations/items would need to be componentized
+* Larger type system needed
+* More structured to keep the amount of types down
+
+Messages would require a more rigid and well defined type structure. For instance for getting credentials from various sources and passing those credentials around we would want to avoid the following.
+
+* MySQLCredentials
+* RedisCredentials
+* RemoteWriteCredentials
+* MySQLCredentialsSourceS3
+* MySQLCredentialsSourceVault
+* MySQLCredentialsSourceConsul
+* RedisCredentialsSourceS3
+* RedisCredentialsSourceVault
+* RedisCredentialsSourceConsul
+* RemoteWriteCredentialsSourceS3
+* RemoteWriteCredentialsSourceVault
+* RemoteWriteCredentialsSourceConsul
+
+And instead have one component that can read from many sources and outputs a single `Credential Type` and its up the destination component to intepret that correctly. 
+
+## 2.3 Hybrid
+
+## 2.4 Examples
+
+### 2.4.1 Simple Example Mysql from Target Discovery
+
+**Expression**
+
+```
+discovery "mysql_pods" {
+    # some sort of config here to find pods
+}
+
+
+integration "mysql" {
+  # Create one mysql integration for every element in the array here 
+  for_each = discovery.mysql_pods.targets
+
+  # Each spawned mysql integration has its data_source_name derived from 
+  # the address label of the input target.
+  data_source_name = "root@(${each.labels["__address__"]})"
+}
+```
+
+**Message**
+
+```
+discovery "mysqlpods" {
+    relabel_config {
+        [
+            {
+                source = "__address__"
+                match = "*mysql"
+                action = "replace"
+                replacement = "root@$1"
+            }
+        ]
+    }
+}
+
+# I think this would depend on convention, mysql would look at __address__ , and maybe optionally look for username/password
+integration "mysql" {}
+
+connections {
+    [
+        {
+            source = mysqlpods
+            destination = mysql
+        }
+    ]
+}
```

Comment:
Is it an artifact of the design that connections between components are defined separate to the components?